Add contentfulness speedindex calculations

dpalmeiro opened the pull request:
Add a new visual metric based on edge detection.  This adds a new option in visualmetrics.py, --contentful, that calculates a speed index using edge detection histograms.  This new metric, designed by Bas Schouten, is an attempt to measure real content on a page load, and heavily favours text.  One big advantage is that it does not suffer from splash screens like the one experienced from tripadvisor.com.  

Additionally, this will remove lastVisualChange and rum speed index from the stdout printing on each iteration, but it will still remain in the browsertime.json output.

ncalexan commented on vendor/visualmetrics.py:
```diff
@@ -1457,6 +1462,41 @@ def calculate_speed_index(progress):
     return int(si)
 
+def calculate_contentful_speed_index(progress, directory):
+    x = len(progress)
```
It looks like `x` is used only once, and then just for indexing the last element.  You can use `progress[-1]` for that in Python.

ncalexan commented on vendor/visualmetrics.py:
```diff
@@ -1457,6 +1462,41 @@ def calculate_speed_index(progress):
+    dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), directory)
+    first_paint_frame = os.path.join(
+        dir, "ms_{0:06d}.png".format(
+            progress[1]["time"]))
```
Indexing starts at 0 in Python.  Do you really want to skip `progress[0]`?  Your subsequent loop runs from `1:`, so you're just dropping the first frame here.

ncalexan commented on vendor/visualmetrics.py:
```diff
@@ -1457,6 +1462,41 @@ def calculate_speed_index(progress):
+        # Full Path of the Current Frame
+        current_frame = os.path.join(dir, "ms_{0:06d}.png".format(p["time"]))
+        logging.debug("Current Image is %s" % current_frame)
+        # Takes full path of PNG frames to compute contentfulness value
```
It would be nice to do everything in parallel here, but that can wait.

ncalexan commented on vendor/visualmetrics.py:
```diff
@@ -1457,6 +1462,41 @@ def calculate_speed_index(progress):
+            progress[1]["time"]))
+    target_frame = os.path.join(
+        dir, "ms_{0:06d}.png".format(progress[x - 1]["time"]))
+    logging.debug("Target image for contSI is %s" % target_frame)
```
Consider not abbreviating and putting the "tag" first, so: "contentfulSpeedIndex: target image is %s".

ncalexan commented on vendor/visualmetrics.py:
```diff
@@ -1457,6 +1462,41 @@ def calculate_speed_index(progress):
+    for p in progress[1:]:
+        # Full Path of the Current Frame
+        current_frame = os.path.join(dir, "ms_{0:06d}.png".format(p["time"]))
+        logging.debug("Current Image is %s" % current_frame)
```
Same: put the "tag" first, so: "contentfulSpeedIndex: current image is %s".

ncalexan commented on vendor/visualmetrics.py:
```diff
@@ -1457,6 +1462,41 @@ def calculate_speed_index(progress):
+        content[i] = float(content[i]) / float(maxContent)
+
+    # Assume 0 content for first frame
+    cont_si = 1 * (progress[1]['time']-progress[0]['time'])
```
nit: it's usual to put spaces between operators, here, in the `xrange` below, in the `i, value` in enumeration loop above, but I just don't care that much for this code.  Up to you.

ncalexan reviewed (approved):
Overall this looks good.  I don't understand why you're removing `rumSpeedIndex`; these little changes will just make it more difficult to upstream things, and to merge upstream.  If you really want to do such things, make them Pre: commits, so that it's always when merges failure that they're "less important" than the meaty commits.

dpalmeiro commented:
Thanks.  I didn't want to clutter up the stdout real estate after each iteration but I can add the rum si back in to ease merging.

dpalmeiro committed:
Put back PerceptualSpeedIndex in the per iteration stduout dumps

dpalmeiro committed:
Add contentfulness speedindex calculations

Add a new visual metric based on edge detection. This adds a new option
in visualmetrics.py, --contentful, that calculates a speed index using
edge detection histograms. This new metric, designed by Bas Schouten, is
an attempt to measure real content on a page load, and heavily favours
text. One big advantage is that it does not suffer from splash screens
like the one experienced from tripadvisor.com.
